feat: add relation between Fid <> Fname entities

gskril opened the pull request:
https://warpcast.com/varunsrin.eth/0x006b00

gskril committed:
feat: add relation between Fid <> Fname entities

typedarray commented on schema.graphql:
```diff
@@ -1,10 +1,12 @@
 type Fid @entity {
   id: BigInt!
   address: String!
+  fnames: [Fname] @derivedFrom(field: "fid")
```
Common GraphQL footgun - the list type and element type should (almost always) both be non-nullable. Though I might be misunderstanding the relationship logic here. https://stackoverflow.com/a/46771890


```gql
fnames: [Fname!]! @derivedFrom(field: "fid")
```

typedarray commented on src/ERC1967Proxy.ts:
```diff
@@ -1,14 +1,27 @@
 ponder.on("ERC1967Proxy:Transfer", async ({ event, context }) => {
   const { Fname } = context.entities;
+
+  const fid = await client.readContract({
```
Ponder natively supports read-only contract calls via the `context.contracts` object. These magical contract instances implement caching to speed up indexing on subsequent hot reloads/redeploys. They also automatically sets the `blockNumber: event.block.number`.  https://ponder.sh/api-reference/event-handlers#readonlycontract

```ts
const fid = await context.contracts.IdRegistry.read.idOf([event.params.to])
```

So, you don't need to create a `client` as you've done here. The provided contract instances are already hooked up to a viem `PublicClient` that Ponder manages internally.

gskril replied:
oh wow this is great, TIL!

typedarray reviewed (commented):
Adding the `fnames` derived field should solve the problem, here are the docs on [one-to-many relationships](https://ponder.sh/guides/design-your-schema#one-to-many-relationships).

The contract call approach seems like it works, but this would be cleaner if there is an event emitted when the Fname/Fid association is first created or updated. Note that it's totally normal to CRUD multiple entities in one event handler.

Note: I don't fully understand the logical relationship between Fnames and Fids so correct me if I'm wrong 😆

gskril committed:
Use ponder's native contract reading

Same functionality but implements caching for performance improvements on restarts

gskril committed:
Fix GraphQL schema

gskril committed:
Add semicolon for the boomers

varunsrin commented on src/ERC1967Proxy.ts:
```diff
@@ -3,12 +3,15 @@ import { bytesToString, toBytes } from "viem";
 ponder.on("ERC1967Proxy:Transfer", async ({ event, context }) => {
   const { Fname } = context.entities;
+
+  const fid = await context.contracts.IdRegistry.read.idOf([event.params.to]);
```
the association between the fid and fname is "broken" if the fid is transferred out to a different address. i don't think this handles that case since it permanently associated the fname with the fid.

gskril replied:
ah good point. just to make sure I follow - are you saying that in the `IdRegistry:Transfer` handler we'd have to update the `fid` of all names owned by the `event.params.from` address?

varunsrin replied:
yep, that's right

gskril replied:
Sounds like we'll have to wait for a ponder update to do this cleanly https://github.com/0xOlias/ponder/issues/317

varunsrin replied:
cc @0xOlias

gskril committed:
Add `Account` entity

Turns out we can just abstract the `fid` <> `fname` relationship through a higher level `Account` entity

typedarray committed:
feat: create relationship from Account to Fid

gskril committed:
Optimize db queries

gskril commented:
Turns out we didn't need new Ponder APIs - a slightly different schema is a cleaner solution. Should be good now!
